upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -1384,5 +1385,67 @@ func (n *nodeUpConfigBuilder) BuildConfig(ig *kops.InstanceGroup, apiserverAddit
 	config.Channels = n.channels
 	config.EtcdManifests = n.etcdManifests[role]
 
+	if ig.Spec.WarmPool != nil || cluster.Spec.WarmPool != nil {
+		config.WarmPoolImages = n.getWarmPoolImages(ig)
+	}
+
 	return config, auxConfig, nil
 }
+
+// getWarmPoolImages returns a list of container images that should be pre-pulled during instance pre-initialization
+func (n *nodeUpConfigBuilder) getWarmPoolImages(ig *kops.InstanceGroup) []string {
+	if ig != nil && ig.Spec.Role != kops.InstanceGroupRoleNode && ig.Spec.Role != kops.InstanceGroupRoleAPIServer {
```

Comment:
Pre-pulling images could make sense for APIServer nodes.

Reply:
Done, I think it's more clear now.